src/connection/host_lookup.c

```diff
@@ -0,0 +1,622 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * host_lookup.c
+ */
+
+#ident "$Id$"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <string.h>
+#include <limits.h>
+#include <errno.h>
+#include <unordered_map>
+
+#if defined (WINDOWS)
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#else
+#include <arpa/inet.h>
+#include <netdb.h>
+#endif
+
+#include "system_parameter.h"
+#include "environment_variable.h"
+#include "message_catalog.h"
+
+#define HOSTNAME_LEN                 (256)
+#define MAX_NUM_HOSTS                (256)
+#define LINE_BUF_SIZE                (512)
+#define IPADDR_LEN                   (17)
+#define NUM_IPADDR_DOT               (3)
+#define IPv4_ADDR_LEN                (4)
+#define USER_HOSTS_FILE              "hosts.conf"
+
+#define NUM_DIGIT(VAL)              (size_t)(log10 (VAL) + 1)
+#define FREE_MEM(PTR)           \
+        do {                    \
+          if (PTR) {            \
+            free(PTR);          \
+            PTR = 0;    \
+          }                     \
+        } while (0)
+
+typedef enum
+{
+  HOSTNAME_TO_IPADDR = 0,
+  IPADDR_TO_HOSTNAME = 1,
+} LOOKUP_TYPE;
+
+typedef enum
+{
+  INSERT_IPADDR = 0,
+  INSERT_HOSTNAME = 1,
+} HOSTENT_INSERT_TYPE;
+
+typedef enum
+{
+  USE_GLIBC_HOSTS = 0,
+  USE_USER_DEFINED_HOSTS = 1,
+} HOST_LOOKUP_TYPE;
+
+typedef enum
+{
+  LOAD_FAIL = -1,
+  LOAD_INIT,
+  LOAD_SUCCESS,
+} HOSTS_CONF_LOAD_STATUS;
+
+static struct hostent *hostent_Cache[MAX_NUM_HOSTS];
+
+static int hosts_conf_file_Load = LOAD_INIT;
+
+// *INDENT-OFF*
+static std::unordered_map <std::string, int> user_host_Map;
+// *INDENT-ON*
+
+static struct hostent *hostent_alloc (char *ipaddr, char *hostname);
+static bool ip_format_check (char *ip_addr);
+static int load_hosts_file ();
+static struct hostent *host_lookup_internal (const char *hostname, struct sockaddr *saddr, LOOKUP_TYPE lookup_type);
+
+/*
+ * hostent_alloc () - Allocate memory hostent structure.
+ * 
+ * return   : the hostent pointer.
+ *
+ */
+static struct hostent *
+hostent_alloc (char *ipaddr, char *hostname)
+{
+  struct hostent *hp;
+  char addr_trans_bi_buf[sizeof (struct in_addr)];
+
+  if ((hp = (struct hostent *) malloc (sizeof (struct hostent))) == NULL)
+    {
+      return NULL;
+    }
+
+  hp->h_addrtype = AF_INET;
+  hp->h_length = IPv4_ADDR_LEN;
+
+  if (inet_pton (AF_INET, ipaddr, addr_trans_bi_buf) < 1)
+    {
+      FREE_MEM (hp);
+      return NULL;
+    }
+
+  hp->h_name = strdup (hostname);
+  hp->h_aliases = NULL;
+
+  if ((hp->h_addr_list = (char **) malloc (sizeof (char *) * MAX_NUM_HOSTS)) == NULL)
```

Comment:
I think you'd better MAX_NUM_IPADDR_PER_HOST at the beginning, for example

#define MAX_NUM_IPADDR_PER_HOST 1

and change this line to:

if ((hp->h_addr_list = (char **) malloc (sizeof (char *) * MAX_NUM_IPADDR_PER_HOST)) == NULL)

Reply:
Ok. I will apply it.